### PR TITLE
Backport GuidanceView to UI SDK

### PR DESCRIPTION
**Pull request by abhishek1508**
## Description

Backport GuidanceView to UI SDK [#249](https://github.com/mapbox/navigation-sdks/issues/249)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Backport GuidanceView to UI SDK

### Implementation

Backport GuidanceView to UI SDK

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: examples/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -32,6 +32,9 @@
     <string name="title_simple_navigation_kotlin" translatable="false">Simple Mapbox Navigation Kotlin</string>
     <string name="description_simple_navigation_kotlin" translatable="false">Shows how to use the Mapbox Navigation in Kotlin code.</string>
 
+    <string name="title_navigate_instruction_view" translatable="false">Navigate with Instruction View</string>
```
Comment (langsmith) [inline examples/src/main/res/values/strings.xml]:
```suggestion
    <string name="title_navigate_instruction_view" translatable="false">Navigate with InstructionView</string>
```

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 2]
```diff
@@ -334,6 +381,33 @@ public void hideInstructionList() {
...
+  private void showGuidanceView(double distanceTraveled, double currentStepTotalDistance) {
+    int thresholdToShowGuidanceView = 200;
+    if (currentStepTotalDistance <= thresholdToShowGuidanceView) {
+      routeJunction = new RouteJunction(appendTokenTo(), currentStepTotalDistance);
+      animateShowGuidanceViewImage();
+      new Picasso.Builder(getContext()).build().load(routeJunction.getGuidanceImageUrl()).into(guidanceViewImage);
+    } else if (distanceTraveled + thresholdToShowGuidanceView >= currentStepTotalDistance) {
+      routeJunction = new RouteJunction(appendTokenTo(), currentStepTotalDistance);
+      new Picasso.Builder(getContext()).build().load(routeJunction.getGuidanceImageUrl()).into(guidanceViewImage);
+      animateShowGuidanceViewImage();
```
Comment (langsmith) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: is there a reason why the order of these two lines is reverse from https://github.com/mapbox/mapbox-navigation-android/pull/2519/files#diff-e993025c73d006405ac10453685867d0R389-R390 ?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 3]
```diff
@@ -334,6 +381,33 @@ public void hideInstructionList() {
     onInstructionListVisibilityChanged(false);
   }
 
+  /////////////////////////////////////////
+  private void showGuidanceView(double distanceTraveled, double currentStepTotalDistance) {
+    int thresholdToShowGuidanceView = 200;
```
Comment (langsmith) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: Add unit into the variable name? 200 what? Meters?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 2 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: Nope. I was trying to see if it helps in animation.

File: examples/src/main/res/values/strings.xml (hunk 1 above)
Comment (abhishek1508) [reply examples/src/main/res/values/strings.xml]: 👍

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java (hunk 3 above)
Comment (abhishek1508) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: That's a good question. I think it should be `meters` irrespective of the distance unit one is following. It should go into the documentation.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/junction/RouteJunction.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package com.mapbox.navigation.ui.junction;
+
+public class RouteJunction {
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/junction/RouteJunction.java]: There's some naming inconsistency, should we use _Guidance_ instead?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/junction/RouteJunctionModel.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
+package com.mapbox.navigation.ui.junction;
+
+import com.mapbox.navigation.base.trip.model.RouteProgress;
+
+public class RouteJunctionModel {
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/junction/RouteJunctionModel.java]: Same as above re: naming inconsistency _Junction_ vs _Guidance_

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 6]
```diff
@@ -239,6 +257,35 @@ public void unsubscribe() {
       navigationViewModel.instructionModel.removeObservers(lifecycleOwner);
       navigationViewModel.bannerInstructionModel.removeObservers(lifecycleOwner);
       navigationViewModel.isOffRoute.removeObservers(lifecycleOwner);
+      navigationViewModel.routeJunctionModel.removeObservers(lifecycleOwner);
+    }
+  }
+
+  ///////////////////////////////////////////////////////////////////
+  public void determineGuidanceView(RouteProgress routeProgress) {
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: IMO `determineGuidanceView` is too long.
What about decomposing / extracting blocks of code into private methods / separate classes with single responsibility and make intention clearer? `InstructionView` is getting cluttered 😬

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java [hunk 7]
```diff
@@ -338,6 +385,32 @@ public void hideInstructionList() {
     onInstructionListVisibilityChanged(false);
   }
 
+  /////////////////////////////////////////
+  private void showGuidanceView(float currentStepDistanceTraveled, double currentStepTotalDistance) {
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/instruction/InstructionView.java]: Same as above re: long method / refactor code out into smaller methods / classes

**Pull request merged**